hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -158,96 +169,63 @@ public Map<DatanodeDetails, SCMCommand<?>> processAndCreateCommands(
+        // issues, but were not able to find any at some stage, and the
+        // placement policy threw an exception.
+        // At this stage. If the cluster is small and there are some
+        // over replicated indexes, it could stop us finding a new node as there
+        // are no more nodes left to try.
+        // If the container is also over replicated, then hand it off to the
+        // over-rep handler, and after those over-rep indexes are cleared the
+        // under replication can be re-tried in the next iteration of RM.
+        // However, we should only hand off to the over rep handler if there are
+        // no commands already created. If we have some commands, they may
+        // attempt to use sources the over-rep handler would remove. So we
+        // should let the commands we have created be processed, and then the
+        // container will be re-processed in a further RM pass.
+        LOG.debug("Unable to located new target nodes for container {}",
+            container, e);
+        if (commands.size() > 0) {
+          LOG.debug("Some commands have already been created, so returning " +
+              "with them only");
+          return commands;
         }
-      }
-      Set<Integer> decomIndexes = replicaCount.decommissioningOnlyIndexes(true);
-      if (decomIndexes.size() > 0) {
-        final List<DatanodeDetails> selectedDatanodes =
-            getTargetDatanodes(excludedNodes, container, decomIndexes.size());
-        if (validatePlacement(nodes, selectedDatanodes)) {
-          excludedNodes.addAll(selectedDatanodes);
-          Iterator<DatanodeDetails> iterator = selectedDatanodes.iterator();
-          // In this case we need to do one to one copy.
-          for (ContainerReplica replica : replicas) {
-            if (decomIndexes.contains(replica.getReplicaIndex())) {
-              if (!iterator.hasNext()) {
-                LOG.warn("Couldn't find enough targets. Available source"
-                    + " nodes: {}, the target nodes: {}, excluded nodes: {} and"
-                    + "  the decommission indexes: {}",
-                    replicas, selectedDatanodes, excludedNodes, decomIndexes);
-                break;
-              }
-              DatanodeDetails decommissioningSrcNode
-                  = replica.getDatanodeDetails();
-              final ReplicateContainerCommand replicateCommand =
-                  new ReplicateContainerCommand(id.getProtobuf().getId(),
-                      ImmutableList.of(decommissioningSrcNode));
-              // For EC containers, we need to track the replica index which is
-              // to be replicated, so add it to the command.
-              replicateCommand.setReplicaIndex(replica.getReplicaIndex());
-              DatanodeDetails target = iterator.next();
-              commands.put(target, replicateCommand);
-            }
-          }
+        if (replicaCount.isOverReplicated()) {
```

Comment:
Should we have this exception handling in UnderReplicatedProcessor & enque over replicated processor queue instead?

Reply:
Wouldn't that be better then only one thread will access overreplication handler. Otherwise underreplication is calling over replication.

Reply:
I guess this could be a valid way of doing it, but I think it just adds more indirection. There is no issue with multiple threads accessing the over-replication handler, infact I think that will happen when the balancer starts using the ReplicationManager methods anyway.

Reply:
In my opinion it is better to not block the under replication handler thread with over replication handling operations. Separating the 2 operations would be a better approach.

Reply:
We can see how this works in practice. I don't expect either of these threads to be a bottleneck in processing, as the work they do is fairly lightweight so I am going to leave this as it is for now.